druid-shell/src/platform/mac/alert.rs

```diff
@@ -0,0 +1,120 @@
+const BUTTON_ID_CANCEL: NSInteger = NSAlertSecondButtonReturn;
+const BUTTON_ID_ALT_START: NSInteger = NSAlertThirdButtonReturn;
+
+pub(crate) unsafe fn show(view: &mut Object, request: AlertRequest) {
+    let alert = NSAlert::alloc(nil).init().autorelease();
+
+    alert.setMessageText(make_nsstring(&request.options.message));
+    alert.setInformativeText(make_nsstring(&request.options.description));
+
+    match request.options.icon {
+        Some(AlertIcon::Information) => alert.setAlertStyle(NSAlertStyle::Informational),
+        Some(AlertIcon::Warning) => alert.setAlertStyle(NSAlertStyle::Warning),
+        Some(AlertIcon::Error) => alert.setAlertStyle(NSAlertStyle::Critical),
+        None => (), // The OS will default to NSAlertStyle::Warning
+    }
+
+    // The primary button is always first. (macOS alert buttons go from right to left)
+    let btn_primary = alert.addButtonWithTitle(make_nsstring(&request.options.primary.label));
+    // Set the primary button as the default which can be chosen via Enter.
+    btn_primary.setKeyEquivalent(make_nsstring("\r"));
+    btn_primary.setTag(BUTTON_ID_PRIMARY);
+
+    // If there's a cancel button, it is always second.
+    if let Some(cancel) = &request.options.cancel {
+        let btn_cancel = alert.addButtonWithTitle(make_nsstring(&cancel.label));
+        // Mark it as the cancel button by adding the Escape hotkey. This also activates Cmd+dot.
+        btn_cancel.setKeyEquivalent(make_nsstring("\u{1b}"));
+        btn_cancel.setTag(BUTTON_ID_CANCEL);
+    }
+
+    // Any alternative buttons are always last.
+    for (idx, alternative) in request.options.alternatives.iter().enumerate() {
+        let btn_alt = alert.addButtonWithTitle(make_nsstring(&alternative.label));
+        // Any hotkeys need to be explicit with custom buttons to avoid duplicates.
+        btn_alt.setKeyEquivalent(make_nsstring(""));
+        btn_alt.setTag(BUTTON_ID_ALT_START + idx as NSInteger);
+    }
+
+    if request.options.app_modal {
+        let tag = alert.runModal();
+        handle_result(view, &request, tag);
+    } else {
+        let window: id = msg_send![view, window];
+        let view_ptr = WeakPtr::new(view as *mut Object);
+
+        let alert_handler = block::ConcreteBlock::new(move |tag: NSInteger| {
+            let view = *view_ptr.load();
+            if !view.is_null() {
+                handle_result(&mut *view, &request, tag);
+            }
+        });
+        let alert_handler = alert_handler.copy(); // Force the block to live on the heap
+
+        alert.beginSheetModalForWindow_completionHandler(window, alert_handler);
+    }
+}
+
+unsafe fn handle_result(view: &mut Object, request: &AlertRequest, tag: NSInteger) {
+    let button = match tag {
+        BUTTON_ID_PRIMARY => Some(request.options.primary.clone()),
+        BUTTON_ID_CANCEL | NSModalResponseAbort | NSModalResponseStop | NSModalResponseCancel => {
+            None
+        }
+        id => {
```

Comment:
I like to include `if` conditions as part of the branch, if possible. I see why that's annoying here, because the conditional is so verbose; one option would be to add a simple function, `fn is_valid_id(id: NSInteger, request: &AlertRequest) -> bool`, so that we can just do the match as `id if is_valid_id(id, request) => { .. }`.

not a blocker or anything, just one of those little refactors that might make this slightly easier to read, not that it's particularly tricky or anything. :)